src/Common/src/Interop/Windows/winhttp/Interop.SafeWinHttpHandle.cs

```diff
@@ -30,6 +30,31 @@ public static void DisposeAndClearHandle(ref SafeWinHttpHandle winHttpHandler)
                 }
             }
 
+            // The SafeHandle base class is inconsistent w.r.t. how Dispose() and DangerousRelease()
+            // are implemented. Both methods can cause the ref count on the SafeHandle to be decremented.
+            // And when the ref count reaches zero, ReleaseHandle() is called and the handle is marked
+            // as 'closed'. However, there is a separate internal state bit to track Dispose() calls.
+            // Calling Dispose() more than once is a no-op.  However, calling DangerousRelease() and then
+            // Dispose() will cause ObjectDisposedException to be thrown if the previous DangerousRelease()
+            // caused the handle to be closed. This seems inconsistent especially since both methods can
+            // decrement the ref count and cause ReleaseHandle() to be called.
```

Comment:
>I'm not understanding this. If you have a dependency that requires that SafeHandle A be alive as long as SafeHandle B is alive, then when you construct B you DangerousAddRef A, and when you destruct B (either Dispose or ~), you DangerousRelease B. There can be any number of "B"s in this scenario: as long as one of them is alive, "A" can't be destroyed. Attempts to Dispose of "A" while there are any "B"s will cause the free'ing of "A" to be delayed until the last DangerousRelease.

In order to implement that, we'll need to have a different design/subclass of the `SafeHandle`.  Right now, there is a single `SafeWinHttpHandle` class.  But it sounds like there needs to a variant of this class that can take a "parent" `SafeHandle` in order to managed AddRef/Release on it as you described. This was not a design I had considered, but sounds interesting and may solve some of these issues.